File: agentlang/lexer.py
from __future__ import annotations

import re
from dataclasses import dataclass


@dataclass(frozen=True)
class Token:
    kind: str
    value: str
    line: int
    col: int


TOKEN_RE = re.compile(
    r"""
    (?P<WS>[ \t\r\n]+)
    |(?P<COMMENT>//[^\n]*)
    |(?P<HASHCOMMENT>\#[^\n]*)
    |(?P<ARROW>->)
    |(?P<EQEQ>==)
    |(?P<NEQ>!=)
    |(?P<STRING>"(?:\\.|[^"\\])*")
    |(?P<NUMBER>\d+(?:\.\d+)?)
    |(?P<LBRACE>\{)
    |(?P<RBRACE>\})
    |(?P<LPAREN>\()
    |(?P<RPAREN>\))
    |(?P<LBRACKET>\[)
    |(?P<RBRACKET>\])
    |(?P<COLON>:)
    |(?P<COMMA>,)
    |(?P<SEMI>;)
    |(?P<EQUAL>=)
    |(?P<PLUS>\+)
    |(?P<DOT>\.)
    |(?P<ID>[A-Za-z_][A-Za-z0-9_]*)
    """,
    re.VERBOSE,
)


KEYWORDS = {
    "agent",
    "task",
    "pipeline",
    "model",
    "tools",
    "let",
    "run",
    "with",
    "by",
    "parallel",
    "join",
    "return",
    "if",
    "else",
    "retries",
    "on_fail",
    "abort",
    "use",
}


class LexError(ValueError):
    pass


def lex(source: str) -> list[Token]:
    tokens: list[Token] = []
    i = 0
    line = 1
    col = 1

    while i < len(source):
        match = TOKEN_RE.match(source, i)
        if not match:
            snippet = source[i : i + 30]
            raise LexError(f"Unexpected token at {line}:{col}: {snippet!r}")

        kind = match.lastgroup
        assert kind is not None
        text = match.group()

        if kind in {"WS", "COMMENT", "HASHCOMMENT"}:
            pass
        elif kind == "ID" and text in KEYWORDS:
            tokens.append(Token(kind=text.upper(), value=text, line=line, col=col))
        elif kind == "STRING":
            tokens.append(Token(kind="STRING", value=_decode_string(text), line=line, col=col))
        elif kind == "NUMBER":
            tokens.append(Token(kind="NUMBER", value=text, line=line, col=col))
        else:
            tokens.append(Token(kind=kind, value=text, line=line, col=col))

        line, col = _advance_position(text, line, col)
        i = match.end()

    tokens.append(Token(kind="EOF", value="", line=line, col=col))
    return tokens


_SIMPLE_ESCAPES: dict[str, str] = {
    "n": "\n",
    "t": "\t",
    "r": "\r",
    "\\": "\\",
    '"': '"',
    "'": "'",
    "0": "\0",
}


def _decode_string(token: str) -> str:
    body = token[1:-1]
    result: list[str] = []
    i = 0
    while i < len(body):
        ch = body[i]
        if ch != "\\":
            result.append(ch)
            i += 1
            continue

        i += 1
        if i >= len(body):
            raise LexError("Unterminated escape sequence in string literal.")

        esc = body[i]
        if esc in _SIMPLE_ESCAPES:
            result.append(_SIMPLE_ESCAPES[esc])
            i += 1
            continue

        if esc == "u":
            if i + 4 >= len(body):
                raise LexError("Incomplete \\uXXXX escape sequence.")
            hex_digits = body[i + 1 : i + 5]
            if not all(c in "0123456789abcdefABCDEF" for c in hex_digits):
                raise LexError(f"Invalid \\uXXXX escape: \\u{hex_digits}")
            result.append(_decode_code_point(hex_digits, "\\uXXXX"))
            i += 5
            continue

        if esc == "U":
            if i + 8 >= len(body):
                raise LexError("Incomplete \\UXXXXXXXX escape sequence.")
            hex_digits = body[i + 1 : i + 9]
            if not all(c in "0123456789abcdefABCDEF" for c in hex_digits):
                raise LexError(f"Invalid \\UXXXXXXXX escape: \\U{hex_digits}")
            result.append(_decode_code_point(hex_digits, "\\UXXXXXXXX"))
            i += 9
            continue

        raise LexError(f"Unknown escape sequence: \\{esc}")

    return "".join(result)


def _decode_code_point(hex_digits: str, form: str) -> str:
    value = int(hex_digits, 16)
    if value > 0x10FFFF or 0xD800 <= value <= 0xDFFF:
        raise LexError(f"Invalid {form} escape: out-of-range code point U+{hex_digits.upper()}")
    return chr(value)


def _advance_position(text: str, line: int, col: int) -> tuple[int, int]:
    parts = text.split("\n")
    if len(parts) == 1:
        return line, col + len(text)
    return line + (len(parts) - 1), len(parts[-1]) + 1

File: src/lexer.rs
use crate::ast::Span;
use thiserror::Error;

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: Kind,
    pub text: String,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Kind {
    Id,
    String,
    Number,
    Symbol,
    Keyword,
    Eof,
}

#[derive(Debug, Error, PartialEq, Eq)]
#[error("[AGL0001] {message} at {line}:{col}")]
pub struct LexError {
    pub message: String,
    pub line: usize,
    pub col: usize,
}

impl LexError {
    pub const fn code(&self) -> &'static str {
        "AGL0001"
    }
}

const KEYWORDS: &[&str] = &[
    "language",
    "agent",
    "tool",
    "task",
    "pipeline",
    "workflow",
    "stage",
    "does",
    "review",
    "checks",
    "revise",
    "using",
    "max_rounds",
    "max_concurrency",
    "model",
    "tools",
    "let",
    "run",
    "with",
    "by",
    "parallel",
    "join",
    "return",
    "if",
    "while",
    "else",
    "break",
    "continue",
    "null",
    "retries",
    "on_fail",
    "abort",
    "use",
    "timeout",
    "true",
    "false",
    "type",
    "enum",
    "try",
    "catch",
    "assert",
    "test",
];

pub fn lex(source: &str) -> Result<Vec<Token>, LexError> {
    let chars: Vec<char> = source.chars().collect();
    let (mut i, mut line, mut col) = (0, 1, 1);
    let mut out = Vec::new();
    while i < chars.len() {
        let ch = chars[i];
        if ch.is_whitespace() {
            advance(ch, &mut line, &mut col);
            i += 1;
            continue;
        }
        if ch == '-' && chars.get(i + 1) == Some(&'-') {
            while i < chars.len() && chars[i] != '\n' {
                i += 1;
                col += 1;
            }
            continue;
        }
        let span = Span { line, col };
        if ch.is_ascii_alphabetic() || ch == '_' {
            let start = i;
            while i < chars.len() && (chars[i].is_ascii_alphanumeric() || chars[i] == '_') {
                i += 1;
                col += 1;
            }
            let text: String = chars[start..i].iter().collect();
            let kind = if KEYWORDS.contains(&text.as_str()) {
                Kind::Keyword
            } else {
                Kind::Id
            };
            out.push(Token { kind, text, span });
            continue;
        }
        if ch.is_ascii_digit() {
            let start = i;
            while i < chars.len() && chars[i].is_ascii_digit() {
                i += 1;
                col += 1;
            }
            if chars.get(i) == Some(&'.') && chars.get(i + 1).is_some_and(|c| c.is_ascii_digit()) {
                i += 1;
                col += 1;
                while i < chars.len() && chars[i].is_ascii_digit() {
                    i += 1;
                    col += 1;
                }
            }
            out.push(Token {
                kind: Kind::Number,
                text: chars[start..i].iter().collect(),
                span,
            });
            continue;
        }
        if ch == '"' {
            i += 1;
            col += 1;
            let mut value = String::new();
            let mut closed = false;
            while i < chars.len() {
                let c = chars[i];
                if c == '"' {
                    i += 1;
                    col += 1;
                    closed = true;
                    break;
                }
                if c == '\n' {
                    return Err(err("newline in string literal", line, col));
                }
                if c != '\\' {
                    value.push(c);
                    i += 1;
                    col += 1;
                    continue;
                }
                i += 1;
                col += 1;
                let esc = *chars
                    .get(i)
                    .ok_or_else(|| err("unterminated escape sequence", line, col))?;
                match esc {
                    'n' => value.push('\n'),
                    't' => value.push('\t'),
                    'r' => value.push('\r'),
                    '\\' => value.push('\\'),
                    '"' => value.push('"'),
                    '\'' => value.push('\''),
                    '0' => value.push('\0'),
                    'u' | 'U' => {
                        let n = if esc == 'u' { 4 } else { 8 };
                        let start = i + 1;
                        let end = start + n;
                        if end > chars.len() {
                            return Err(err("incomplete unicode escape", line, col));
                        }
                        let hex: String = chars[start..end].iter().collect();
                        let cp = u32::from_str_radix(&hex, 16)
                            .map_err(|_| err("invalid unicode escape", line, col))?;
                        value.push(
                            char::from_u32(cp)
                                .ok_or_else(|| err("invalid unicode code point", line, col))?,
                        );
                        i = end;
                        col += n + 1;
                        continue;
                    }
                    _ => return Err(err(&format!("unknown escape sequence \\{esc}"), line, col)),
                }
                i += 1;
                col += 1;
            }
            if !closed {
                return Err(err("unterminated string literal", span.line, span.col));
            }
            out.push(Token {
                kind: Kind::String,
                text: value,
                span,
            });
            continue;
        }
        let two = chars.get(i + 1).map(|b| format!("{ch}{b}"));
        if matches!(two.as_deref(), Some("->" | "==" | "!=")) {
            out.push(Token {
                kind: Kind::Symbol,
                text: two.unwrap(),
                span,
            });
            i += 2;
            col += 2;
            continue;
        }
        if "{}()[]:,;=+.".contains(ch) {
            out.push(Token {
                kind: Kind::Symbol,
                text: ch.to_string(),
                span,
            });
            i += 1;
            col += 1;
            continue;
        }
        return Err(err(&format!("unexpected character {ch:?}"), line, col));
    }
    out.push(Token {
        kind: Kind::Eof,
        text: String::new(),
        span: Span { line, col },
    });
    Ok(out)
}

fn advance(ch: char, line: &mut usize, col: &mut usize) {
    if ch == '\n' {
        *line += 1;
        *col = 1
    } else {
        *col += 1
    }
}
fn err(message: &str, line: usize, col: usize) -> LexError {
    LexError {
        message: message.into(),
        line,
        col,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn comments_and_unicode_strings() {
        let t = lex("-- hi\ntask x() -> String; \"hi\\u0021\"").unwrap();
        assert_eq!(t[0].text, "task");
        assert_eq!(t[7].text, "hi!");
    }
    #[test]
    fn reports_bad_escape() {
        assert!(
            lex("\"\\q\"")
                .unwrap_err()
                .to_string()
                .contains("unknown escape")
        );
    }
}
